### DISCUSSION TITLE
Track parity of require("util") with node

### DISCUSSION THREAD

**Discussion by soneymathew** [oven-sh/bun#852] [General]
To track parity of bun's util with [Node's Util](https://nodejs.org/api/util.html#util) with sufficient tests.

Legend
- API - refers to API in node's util
- Status - Current implementation status in Bun
  - _builtin_ - denotes implementation backed by Zig/C++
  - _fallback_ - denotes implementation backed by Javascript
  - _unsupported_ - Support yet to be added
- Deprecated - denotes if the API is deprecated currently in Node
- Typed - Are the APIs typed and added to `[util.d.ts](https://github.com/oven-sh/bun-types/blob/master/util.d.ts)`
- ❓ - /to-be-validated
- ❌ - not-done
- ✅  - done

| API | Status | Tests | Typed | Deprecated 
|--|:--:|:--:|:--:|:--:|
[util.callbackify(original)](https://nodejs.org/api/util.html#utilcallbackifyoriginal) | _builtin_ | ❌  | ❓| N
[util.debuglog(section[, callback])](https://nodejs.org/api/util.html#utildebuglogsection-callback) | _builtin_ | ❌  | ❓| N
⮑[debuglog().enabled](https://nodejs.org/api/util.html#debuglogenabled) | _unsupported_ | ❌  | ❓| N
[util.debug(section)](https://nodejs.org/api/util.html#utildebugsection) | _unsupported_ | ❌  | ❓| N
[util.deprecate(fn, msg[, code])](https://nodejs.org/api/util.html#utildeprecatefn-msg-code) | _builtin_ | ❌  | ❓| N
[util.format(format[, ...args])](https://nodejs.org/api/util.html#utilformatformat-args) | _builtin_ | ❌  | ❓| N
[util.formatWithOptions(inspectOptions, format[, ...args])](https://nodejs.org/api/util.html#utilformatwithoptionsinspectoptions-format-args) | _unsupported_ | ❌  | ❓| N
[util.getSystemErrorName(err)](https://nodejs.org/api/util.html#utilgetsystemerrornameerr) | _unsupported_ | ❌  | ❓| N
[util.getSystemErrorMap()](https://nodejs.org/api/util.html#utilgetsystemerrormap) | _unsupported_ | ❌  | ❓| N
[util.inherits(constructor, superConstructor)](https://nodejs.org/api/util.html#utilinheritsconstructor-superconstructor) | _builtin_ | ❌  | ❓| N
[util.inspect(object[, options])](https://nodejs.org/api/util.html#utilinspectobject-options) | _builtin_ | ❌  | ❓| N
[util.inspect(object[, showHidden[, depth[, colors]]])](https://nodejs.org/api/util.html#utilinspectobject-showhidden-depth-colors) | _builtin_ | ❌  | ❓| N
⮑[Customizing util.inspect colors](https://nodejs.org/api/util.html#customizing-utilinspect-colors) | _unsupported_ | ❌  | ❓| N
‎↳⟶[Modifiers](https://nodejs.org/api/util.html#modifiers) | _unsupported_ | ❌  | ❓| N
⮑⟶[Foreground colors](https://nodejs.org/api/util.html#foreground-colors) | _unsupported_ | ❌  | ❓| N
⮑⟶[Background colors](https://nodejs.org/api/util.html#background-colors) | _unsupported_ | ❌  | ❓| N
⮑[Custom inspection functions on objects](https://nodejs.org/api/util.html#custom-inspection-functions-on-objects) | _unsupported_ | ❌  | ❓| N
⮑[util.inspect.custom](https://nodejs.org/api/util.html#utilinspectcustom) | _unsupported_ | ❌  | ❓| N
⮑[util.inspect.defaultOptions](https://nodejs.org/api/util.html#utilinspectdefaultoptions) | _unsupported_ | ❌  | ❓| N
[util.isDeepStrictEqual(val1, val2)](https://nodejs.org/api/util.html#utilisdeepstrictequalval1-val2) | _unsupported_ | ❌  | ❓| N
[util.parseArgs([config])](https://nodejs.org/api/util.html#utilparseargsconfig) | _unsupported_ | ❌  | ❓| N
[util.promisify(original)](https://nodejs.org/api/util.html#utilpromisifyoriginal) | _builtin_ | ❌  | ❓| N
⮑[Custom promisified functions](https://nodejs.org/api/util.html#custom-promisified-functions) | _unsupported_ | ❌  | ❓| N
⮑[util.promisify.custom](https://nodejs.org/api/util.html#utilpromisifycustom) | _unsupported_ | ❌  | ❓| N
[util.stripVTControlCharacters(str)](https://nodejs.org/api/util.html#utilstripvtcontrolcharactersstr) | _unsupported_ | ❌  | ❓| N
[Class: util.TextDecoder](https://nodejs.org/api/util.html#class-utiltextdecoder) | _unsupported_ | ❌  | ❓| N
⮑[WHATWG supported encodings](https://nodejs.org/api/util.html#whatwg-supported-encodings) | _unsupported_ | ❌  | ❓| N
⮑⟶[Encodings supported by default (with full ICU data)](https://nodejs.org/api/util.html#encodings-supported-by-default-with-full-icu-data) | _unsupported_ | ❌  | ❓| N
⮑⟶[Encodings supported when Node.js is built with the small-icu option](https://nodejs.org/api/util.html#encodings-supported-when-nodejs-is-built-with-the-small-icu-option) | _unsupported_ | ❌  | ❓| N
⮑⟶[Encodings supported when ICU is disabled](https://nodejs.org/api/util.html#encodings-supported-when-icu-is-disabled) | _unsupported_ | ❌  | ❓| N
⮑[new TextDecoder([encoding[, options]])](https://nodejs.org/api/util.html#new-textdecoderencoding-options)| _unsupported_ | ❌  | ❓| N
⮑[textDecoder.decode([input[, options]])](https://nodejs.org/api/util.html#textdecoderdecodeinput-options) | _unsupported_ | ❌  | ❓| N
⮑[textDecoder.encoding](https://nodejs.org/api/util.html#textdecoderencoding) | _unsupported_ | ❌  | ❓| N
⮑[textDecoder.fatal](https://nodejs.org/api/util.html#textdecoderfatal) | _unsupported_ | ❌  | ❓| N
⮑[textDecoder.ignoreBOM](https://nodejs.org/api/util.html#textdecoderignorebom) | _unsupported_ | ❌  | ❓| N
[Class: util.TextEncoder](https://nodejs.org/api/util.html#class-utiltextencoder) | _builtin_ | ✅   | ❓| N
⮑[textEncoder.encode([input])](https://nodejs.org/api/util.html#textencoderencodeinput) | _builtin_ | ❌  | ❓| N
⮑[textEncoder.encodeInto(src, dest)](https://nodejs.org/api/util.html#textencoderencodeintosrc-dest) | _builtin_ | ❌  | ❓| N
⮑[textEncoder.encoding](https://nodejs.org/api/util.html#textencoderencoding) | _builtin_ | ❌  | ❓| N
[util.toUSVString(string)](https://nodejs.org/api/util.html#utiltousvstringstring) | _unsupported_ | ❌  | ❓| N
[util.types](https://nodejs.org/api/util.html#utiltypes) | _builtin_ | ❌  | ❓| N
⮑[util.types.isAnyArrayBuffer(value)](https://nodejs.org/api/util.html#utiltypesisanyarraybuffervalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isArrayBufferView(value)](https://nodejs.org/api/util.html#utiltypesisarraybufferviewvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isArgumentsObject(value)](https://nodejs.org/api/util.html#utiltypesisargumentsobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isArrayBuffer(value)](https://nodejs.org/api/util.html#utiltypesisarraybuffervalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isAsyncFunction(value)](https://nodejs.org/api/util.html#utiltypesisasyncfunctionvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isBigInt64Array(value)](https://nodejs.org/api/util.html#utiltypesisbigint64arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isBigUint64Array(value)](https://nodejs.org/api/util.html#utiltypesisbiguint64arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isBooleanObject(value)](https://nodejs.org/api/util.html#utiltypesisbooleanobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isBoxedPrimitive(value)](https://nodejs.org/api/util.html#utiltypesisboxedprimitivevalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isCryptoKey(value)](https://nodejs.org/api/util.html#utiltypesiscryptokeyvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isDataView(value)](https://nodejs.org/api/util.html#utiltypesisdataviewvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isDate(value)](https://nodejs.org/api/util.html#utiltypesisdatevalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isExternal(value)](https://nodejs.org/api/util.html#utiltypesisexternalvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isFloat32Array(value)](https://nodejs.org/api/util.html#utiltypesisfloat32arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isFloat64Array(value)](https://nodejs.org/api/util.html#utiltypesisfloat64arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isGeneratorFunction(value)](https://nodejs.org/api/util.html#utiltypesisgeneratorfunctionvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isGeneratorObject(value)](https://nodejs.org/api/util.html#utiltypesisgeneratorobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isInt8Array(value)](https://nodejs.org/api/util.html#utiltypesisint8arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isInt16Array(value)](https://nodejs.org/api/util.html#utiltypesisint16arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isInt32Array(value)](https://nodejs.org/api/util.html#utiltypesisint32arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isKeyObject(value)](https://nodejs.org/api/util.html#utiltypesiskeyobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isMap(value)](https://nodejs.org/api/util.html#utiltypesismapvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isMapIterator(value)](https://nodejs.org/api/util.html#utiltypesismapiteratorvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isModuleNamespaceObject(value)](https://nodejs.org/api/util.html#utiltypesismodulenamespaceobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isNativeError(value)](https://nodejs.org/api/util.html#utiltypesisnativeerrorvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isNumberObject(value)](https://nodejs.org/api/util.html#utiltypesisnumberobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isPromise(value)](https://nodejs.org/api/util.html#utiltypesispromisevalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isProxy(value)](https://nodejs.org/api/util.html#utiltypesisproxyvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isRegExp(value)](https://nodejs.org/api/util.html#utiltypesisregexpvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isSet(value)](https://nodejs.org/api/util.html#utiltypesissetvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isSetIterator(value)](https://nodejs.org/api/util.html#utiltypesissetiteratorvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isSharedArrayBuffer(value)](https://nodejs.org/api/util.html#utiltypesissharedarraybuffervalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isStringObject(value)](https://nodejs.org/api/util.html#utiltypesisstringobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isSymbolObject(value)](https://nodejs.org/api/util.html#utiltypesissymbolobjectvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isTypedArray(value)](https://nodejs.org/api/util.html#utiltypesistypedarrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isUint8Array(value)](https://nodejs.org/api/util.html#utiltypesisuint8arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isUint8ClampedArray(value)](https://nodejs.org/api/util.html#utiltypesisuint8clampedarrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isUint16Array(value)](https://nodejs.org/api/util.html#utiltypesisuint16arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isUint32Array(value)](https://nodejs.org/api/util.html#utiltypesisuint32arrayvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isWeakMap(value)](https://nodejs.org/api/util.html#utiltypesisweakmapvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isWeakSet(value)](https://nodejs.org/api/util.html#utiltypesisweaksetvalue) | _builtin_ | ❌  | ❓| N
⮑[util.types.isWebAssemblyCompiledModule(value)](https://nodejs.org/api/util.html#utiltypesiswebassemblycompiledmodulevalue)| _builtin_ | ❌  | ❓| Y
[util._extend(target, source)](https://nodejs.org/api/util.html#util_extendtarget-source) | _unsupported_ | ❌  | ❓| Y
[util.isArray(object)](https://nodejs.org/api/util.html#utilisarrayobject) | _unsupported_ | ❌  | ❓| Y
[util.isBoolean(object)](https://nodejs.org/api/util.html#utilisbooleanobject) | _unsupported_ | ❌  | ❓| Y
[util.isBuffer(object)](https://nodejs.org/api/util.html#utilisbufferobject) | _unsupported_ | ❌  | ❓| Y
[util.isDate(object)](https://nodejs.org/api/util.html#utilisdateobject) | _unsupported_ | ❌  | ❓| Y
[util.isError(object)](https://nodejs.org/api/util.html#utiliserrorobject) | _unsupported_ | ❌  | ❓| Y
[util.isFunction(object)](https://nodejs.org/api/util.html#utilisfunctionobject) | _unsupported_ | ❌  | ❓| Y
[util.isNull(object)](https://nodejs.org/api/util.html#utilisnullobject) | _unsupported_ | ❌  | ❓| Y
[util.isNullOrUndefined(object)](https://nodejs.org/api/util.html#utilisnullorundefinedobject) | _unsupported_ | ❌  | ❓| Y
[util.isNumber(object)](https://nodejs.org/api/util.html#utilisnumberobject) | _unsupported_ | ❌  | ❓| Y
[util.isObject(object)](https://nodejs.org/api/util.html#utilisobjectobject) | _unsupported_ | ❌  | ❓| Y
[util.isPrimitive(object)](https://nodejs.org/api/util.html#utilisprimitiveobject) | _unsupported_ | ❌  | ❓| Y
[util.isRegExp(object)](https://nodejs.org/api/util.html#utilisregexpobject) | _unsupported_ | ❌  | ❓| Y
[util.isString(object)](https://nodejs.org/api/util.html#utilisstringobject) | _unsupported_ | ❌  | ❓| Y
[util.isSymbol(object)](https://nodejs.org/api/util.html#utilissymbolobject) | _unsupported_ | ❌  | ❓| Y
[util.isUndefined(object)](https://nodejs.org/api/util.html#utilisundefinedobject) | _unsupported_ | ❌  | ❓| Y
[util.log(string)](https://nodejs.org/api/util.html#utillogstring) | _unsupported_ | ❌  | ❓| Y

**Comment by albertpurnama** [comment]
I'd love to help to tackle these tests. 

How big should the PR size be? I'm thinking maybe I can start with testing 1-2 functions (<50 line changes), open PR, then after accepted & merged, move on to more functions per PR.

This is to make sure I'm implementing the tests correctly. What do you think?

**Reply by soneymathew** [reply]
@albertputrapurnama One suggestion from my side is surface the gaps with node parity first.
In my first iteration I commented out the tests that are not testable until more zig/C++ changes can be made.
PR size can depend on what you are comfortable with 🙏
Consider mapping the file name to follow similar convention as used by node. so that we can review against node's tests to validate parity.

For example: The tests in [util.test.js](https://github.com/oven-sh/bun/blob/main/test/bun.js/util.test.js) directly maps to [test-util.js](https://github.com/nodejs/node/blob/main/test/parallel/test-util.js) 
perhaps we can move it to a `util` folder ?
and follow a convention like 
test/bun.js/util/test-util.test.js
test/bun.js/util/<filename.in.node.repo>.test.js

**Reply by albertpurnama** [reply]
I'm not sure what the test convention in Node is, I can see multiple different types of test for util inside the `test/parallel` directory of node.

I think following `test/bun.js/util/<filename-without-test-prefix>.test.js` should be ok. I'll try to test https://github.com/nodejs/node/blob/main/test/parallel/test-util-promisify.js using bun, also implement the folder convention as you suggested!

I'll open a PR soon!

**Reply by albertpurnama** [reply]
https://github.com/oven-sh/bun/pull/1018#discussion_r939714883

@soneymathew Do you know anything about `common.mustCall`? I tried reading the code on what it does, I'm having a hard time understanding what it does and whether we should implement these common test utilities on bun. Do you have any thoughts/suggestions on this?

**Comment by albertpurnama** [comment]
`util.types` tests implemented https://github.com/oven-sh/bun/pull/1278
